cmd/run.go

```diff
@@ -0,0 +1,289 @@
+package cmd
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"path/filepath"
+	"runtime"
+	"time"
+
+	"github.com/DataDog/temporalite"
+	"github.com/google/uuid"
+	"github.com/pmezard/go-difflib/difflib"
+	"github.com/urfave/cli/v2"
+	"go.temporal.io/sdk-features/harness/go/cmd"
+	"go.temporal.io/sdk-features/harness/go/history"
+	"go.temporal.io/sdk/client"
+	"go.temporal.io/server/common/log"
+	"go.temporal.io/server/common/log/tag"
+)
+
+func runCmd() *cli.Command {
+	var config RunConfig
+	return &cli.Command{
+		Name:  "run",
+		Usage: "run a test or set of tests",
+		Flags: config.flags(),
+		Action: func(ctx *cli.Context) error {
+			return NewRunner(config).Run(ctx.Context, ctx.Args().Slice())
+		},
+	}
+}
+
+// RunConfig is configuration for NewRunner.
+type RunConfig struct {
+	Lang                string
+	Version             string
+	GenerateHistory     bool
+	DisableHistoryCheck bool
+	// If not set, defaults to creating one
+	Server string
+	// Defaults to unique value
+	Namespace     string
+	RetainTempDir bool
+}
+
+func (r *RunConfig) flags() []cli.Flag {
+	return []cli.Flag{
+		&cli.StringFlag{
+			Name:        "lang",
+			Usage:       "SDK language to run ('go' or 'java' or 'ts')",
+			Required:    true,
+			Destination: &r.Lang,
+		},
+		&cli.StringFlag{
+			Name:        "version",
+			Usage:       "SDK language version to run",
+			Destination: &r.Version,
+		},
+		&cli.BoolFlag{
+			Name:        "generate-history",
+			Usage:       "Generate the history of the features that are run",
+			Destination: &r.GenerateHistory,
+		},
+		&cli.BoolFlag{
+			Name:        "no-history-check",
+			Usage:       "Do not verify history matches",
+			Destination: &r.DisableHistoryCheck,
+		},
+		&cli.StringFlag{
+			Name:        "server",
+			Usage:       "The host:port of the server (default is to create ephemeral in-memory server)",
+			Destination: &r.Server,
+		},
+		&cli.StringFlag{
+			Name:        "namespace",
+			Usage:       "The namespace to use (default is random)",
+			Destination: &r.Namespace,
+		},
+		&cli.BoolFlag{
+			Name:        "retain-temp-dir",
+			Usage:       "Do not delete the temp directory after the run",
+			Destination: &r.RetainTempDir,
+		},
+	}
+}
+
+// Runner can run features.
+type Runner struct {
+	log    log.Logger
+	config RunConfig
+	// Root of the sdk-features repo
+	rootDir    string
+	createTime time.Time
+}
+
+// NewRunner creates a new runner for the given config.
+func NewRunner(config RunConfig) *Runner {
+	return &Runner{
+		// TODO(cretz): Configurable logger
+		log:        log.NewCLILogger(),
+		config:     config,
+		rootDir:    rootDir(),
+		createTime: time.Now(),
+	}
+}
+
+// Run runs all matching features for the given patterns (or all if no patterns
+// given).
+func (r *Runner) Run(ctx context.Context, patterns []string) error {
+	if r.config.Lang != "go" && r.config.Lang != "java" && r.config.Lang != "ts" {
```

Comment:
nit: Consider lifting into list

Reply:
I will if/when used in another place. Go doesn't have a generic "contains" method that makes a list worth it (I can have a empty-value map w/ these as keys, but meh).

Reply:
Oh. Right. Go. :face_exhaling: